commcare_connect/events/tasks.py

```diff
@@ -0,0 +1,37 @@
+import pickle
+from datetime import datetime
+
+from django.db import transaction
+from django_redis import get_redis_connection
+
+from config import celery_app
+
+from .models import Event
+
+REDIS_EVENTS_QUEUE = "events_queue"
+
+
+@celery_app.task
+def process_events_batch():
```

Comment:
Thanks for describing your thought process a bit. A few questions
1) Why redis instead of postgres for the queue?
2) Why does this need to be async at all? I can imagine its slightly faster to write to redis than postgres, but if we are doing a network write for every event anyway, I am surprised this is much more efficient than just writing directly to postgres?

Reply:
Main reason being: Wouldn't writing to redis (in memory) be faster than writing to Postgres DB (on disk)?

Reply:
I would rather we start with a purely synchronous implementation for now if we want to be sure about event durability. We can always make these async later, but this adds a lot of complexity for something we aren't sure we need yet. If we do want to make it  async, I think we should use celery unless we have a very strong reason not to since we are already using that. That keeps us to only one async task method in our code and makes it a bit easier to manage.

Reply:
I am pretty sure that it would impact performance in sync mode, at least in some cases if not all. For e.g. in `process_learn_form` we need to track multiple events (Finishes a Module, Finishes all Modules, Failed Assessment, Succeeded Assessment). Without async, these will add up enough to slow down the request. 

Though, this isn't a big issue if these requests are coming from Repeaters as opposed to directly from mobile submissions.

I am happy to do it in sync mode, but ¯\_(ツ)_/,  you will have to handle if/when things slow down :turtle: :hourglass: :snail:

Reply:
My instinct was that nearly all of these come from background tasks (like approvals) or non time sensitive requests. It is certainly possible that future ones won't (the mobile endpoint should be performant, but we also need to be able to tell mobile whether the write succeeded and we can't do that with an async task anyway), and we can deal with async then, but at this point it seems more important to have guarantees. I am also less convinced of the long term sustainability of this implementation. My understanding of the code is that if _any_ event has an issue, no events will ever be written, based on the error handling there. For example if mobile sends a new event and we havent yet migrated, or a web dev forgets to add the migration. That seems quite dangerous. Additionally its not paged at all, if the list is very long (maybe mobile sends a bunch at once), it could consume a lot of memory.

None of these issues are unfixable, but the more we work to make this a production ready queuing system, the more complex it becomes, and the more we benefit from a real one. However, given our current needs, unless there is a clear place where the async will benefit us, it is a bit hard for me to believe that one postgres write will be enough worse than the one redis write at this time and for our use cases.

I am not going to block the PR over this, and if you think there is a specific place where that performance matters (the form processor would be reasonable if that were blocking a mobile call), I am certainly willing to consider it, but my strong presumption is that this is overkill for our current needs, while not being robust enough to be a long term solution.